### DISCUSSION TITLE
Using type stubs through mypy installed via PDM?

### DISCUSSION THREAD

**Discussion by twoev** [pdm-project/pdm#751] [Q&A]
Hi,

We are using PDM version 1.8.1 with Python 3.7 (we have also experimented with Python 3.9, where we have the same issue).  We are also using Mypy, installed and steered through PDM, to type check our Python code.

The problem we are having starts with Mypy being unable to locate the type stub packages for external libraries, for example `jsonschema` needs an additional `types-jsonschema` package to be installed.  We have added these type stub packages with pdm, and confirm they are present in the `__pypackages__` structure.  In order to find type stubs installed by pdm, my understanding is that the `mypy_path` config option needs to be added to Mypy.  On adding this to the `[tool.mypy]` section of the `pyproject.toml` file Mypy does seem to include the pdm installations in its search path, but unfortunately mypy now fails with the error

```
mypy: "__pypackages__/3.9/lib/typing_extensions.py" shadows library module "typing_extensions"
note: A user-defined top-level module with name "typing_extensions" is not supported
```

The `typing_extensions` module is in fact required by Mypy and gets installed by PDM as part of the Mypy installation.  Mypy then finds its own dependency and (incorrectly) assumes there is a name clash because it thinks the `typing_extensions` module it found is a user module.

The general problem seems to be that Mypy expects to be installed globally and doesn't play nicely with the local PDM installation inside `__pypackages__`.  There is a proposed patch to Mypy for this, but it seems like the devs are not going to accept it

https://github.com/python/mypy/pull/10833

I'm hoping the PDM experts have experienced this issue and maybe have a way around it for the time being?

Thanks for your help

James

**Comment by pawamoy** [comment]
Related mypy feature request: https://github.com/python/mypy/issues/10633

**Reply by twoev** [reply]
Thanks. Does this mean there is nothing that can be done until mypy is fixed?  Like you I have tried making symlinks and pointing `my_path` at the symlinks, but it doesn't really work.

**Reply by pawamoy** [reply]
In fact it does work, I was not using the right variable name: `MYPYPATH`. I've posted [some Python code](https://github.com/python/mypy/issues/10633#issuecomment-974840203) to automate it before mypy runs.

**Reply by twoev** [reply]
Instead of using the `MYPYPATH` environment variable, we set `mypy_path` in the `tool.mypy` section.  Mypy is responding to that config change because we can see its behaviour changes.  Unfortunately, with the symlinks mypy seems unable to reliably follow all the import paths and complains about a bunch of missing definitions in standard packages.

**Reply by pawamoy** [reply]
Ah yes, I have to ignore the errors for packages that are not mine indeed. Well, this is definitely a hack 😅 
We'll have to wait on mypy to bring proper support 😕
